pkg/prowloader/github/github.go

```diff
@@ -39,39 +46,73 @@ func New(ctx context.Context) *Client {
 		}
 	}
 
-	if token != "" {
-		ts := oauth2.StaticTokenSource(
-			&oauth2.Token{
-				AccessToken: token,
-			},
-		)
-		tc := oauth2.NewClient(client.ctx, ts)
-		client.client = gh.NewClient(tc)
-	} else {
-		logrus.Warningf("using unathenticated GitHub client, requests will be rate-limited")
-		client.client = gh.NewClient(nil)
+	if PRFetch != nil {
+		var ghc *gh.Client
+
+		if token != "" {
+			ts := oauth2.StaticTokenSource(
+				&oauth2.Token{
+					AccessToken: token,
+				},
+			)
+			tc := oauth2.NewClient(client.ctx, ts)
+			ghc = gh.NewClient(tc)
+		} else {
+			logrus.Warningf("using unathenticated GitHub client, requests will be rate-limited")
+			ghc = gh.NewClient(nil)
+		}
+
+		PRFetch = func(org, repo string, number int) (*gh.PullRequest, error) {
+			pr, _, err := ghc.PullRequests.Get(client.ctx, org, repo, number)
+			return pr, err
+		}
 	}
 
 	return client
 }
 
-func (c *Client) GetPRMerged(org, repo string, number int, sha string) (*time.Time, error) {
-	prl := prlocator{org: org, repo: repo, number: number, sha: sha}
-	if val, ok := c.cache[prl]; ok {
-		return val, nil
+// GetPRSHAMerged returns the merge time for a PR/SHA combination. The caching is designed
+// to minimize queries to GitHub. We basically have to handle these cases:
```

Comment:
Did you consider just using the db as the cache? Too slow perhaps? (individual queries per call here)

The cache looks unbounded, it would grow forever as long as sippy was running as new PRs come and go. Surely we'd restart often enough to prevent an OOM though so probably good for now. We could expire someday if it becomes a problem.

Reply:
We only use this with --load-database so it's lifetime is limited

Reply:
The DB also has the old state, not the new state, so using it as a cache would be tough.